ui-speedspacechart/src/components/helpers/drawElements/curve.ts

```diff
@@ -0,0 +1,40 @@
+import { clearCanvas, maxPositionValues, speedRangeValues } from '../../utils';
+import type { Store } from '../../../types/chartTypes';
+import { MARGINS } from '../../const';
+
+const { CURVE_MARGIN_TOP, CURVE_MARGIN_SIDES } = MARGINS;
+
+export const drawCurve = (
+  ctx: CanvasRenderingContext2D,
+  width: number,
+  height: number,
+  store: Store
+) => {
+  clearCanvas(ctx, width, height);
+  ctx.translate(store.leftOffset, 0);
+
+  const { minSpeed, speedRange } = speedRangeValues(store);
+  const { maxPosition } = maxPositionValues(store);
+
+  ctx.fillStyle = 'rgba(17, 101, 180, 0.02)';
+  ctx.strokeStyle = 'rgb(0, 0, 0)';
+  ctx.lineWidth = 0.5;
+
+  ctx.beginPath();
+  store.speed.forEach((data) => {
+    // normalize speed based on range of values
+    const normalizedSpeed = (data.speed - minSpeed) / speedRange;
+    const x =
+      data.position * ((width - CURVE_MARGIN_SIDES) / maxPosition) * store.ratioX +
+      CURVE_MARGIN_SIDES / 2;
+    const y = height - normalizedSpeed * (height - CURVE_MARGIN_TOP);
+    ctx.lineTo(x, y);
+  });
+
+  // add fill() before stroke() to avoid overlapping while filling the area
+  ctx.fill();
+
+  ctx.stroke();
+
+  ctx.translate(-store.leftOffset, 0);
```

Comment:
Here is the whole thing:

```suggestion
  ctx.save();
  ctx.translate(store.leftOffset, 0);
  ctx.translate(CURVE_MARGIN_SIDES / 2, 0);
  ctx.scale((width - CURVE_MARGIN_SIDES) / width, 1);
  const { minSpeed, speedRange } = speedRangeValues(store);
  const { maxPosition } = maxPositionValues(store);

  ctx.fillStyle = 'rgba(17, 101, 180, 0.02)';
  ctx.strokeStyle = 'rgb(0, 0, 0)';
  ctx.lineWidth = 0.5;

  ctx.beginPath();
  store.speed.forEach((data) => {
    // normalize speed based on range of values
    const normalizedSpeed = (data.speed - minSpeed) / speedRange;
    const x = data.position * (width / maxPosition) * store.ratioX;
    const y = height - normalizedSpeed * (height - CURVE_MARGIN_TOP);
    ctx.lineTo(x, y);
  });

  // add fill() before stroke() to avoid overlapping while filling the area
  ctx.fill();

  ctx.stroke();
  ctx.restore();
```

Reply:
Can you just do the `save` and `restore` for this PR ?

```suggestion
  ctx.save();
  ctx.translate(store.leftOffset, 0);

  const { minSpeed, speedRange } = speedRangeValues(store);
  const { maxPosition } = maxPositionValues(store);

  ctx.fillStyle = 'rgba(17, 101, 180, 0.02)';
  ctx.strokeStyle = 'rgb(0, 0, 0)';
  ctx.lineWidth = 0.5;

  ctx.beginPath();
  store.speed.forEach((data) => {
    // normalize speed based on range of values
    const normalizedSpeed = (data.speed - minSpeed) / speedRange;
    const x =
      data.position * ((width - CURVE_MARGIN_SIDES) / maxPosition) * store.ratioX +
      CURVE_MARGIN_SIDES / 2;
    const y = height - normalizedSpeed * (height - CURVE_MARGIN_TOP);
    ctx.lineTo(x, y);
  });

  // add fill() before stroke() to avoid overlapping while filling the area
  ctx.fill();

  ctx.stroke();

  ctx.restore();
```

Reply:
of course, done